workflow-api/src/lib.rs

```diff
@@ -0,0 +1,205 @@
+    fn queries() -> &'static [&'static QueryDefinition<Self>] {
+        // TODO
+        &[]
+    }
+}
+
+/// TODO: Exists in SDK in slightly different form, and would move into this crate
+#[derive(Debug)]
+pub enum WfExitValue<T: TemporalSerializable> {
+    /// Continue the workflow as a new execution
+    ContinueAsNew(Box<ContinueAsNewWorkflowExecution>), // Wouldn't be raw proto in reality
+    /// Confirm the workflow was cancelled
+    Cancelled,
+    /// Finish with a result
+    Normal(T),
+}
+impl<T: TemporalSerializable> From<T> for WfExitValue<T> {
+    fn from(v: T) -> Self {
+        Self::Normal(v)
+    }
+}
+// ... also convenience functions for constructing C-A-N, etc.
+
+/// A workflow context which contains only information, but does not allow any commands to
+/// be created.
+pub struct SafeWfContext {
+    // TODO
+}
+
+/// TODO: Placeholder, exists in SDK and would move into this crate & (likely) become a trait
+pub struct WfContext {}
+impl WfContext {
+    pub async fn timer(&self, _: Duration) {
+        todo!()
+    }
+}
+
+pub struct SignalDefinition<WF: Workflow> {
+    // TODO: Could be a matching predicate, to allow for dynamic registration
+    name: String,
+    // The handler input type must be erased here, since otherwise we couldn't store/return the
+    // heterogeneous collection of definition types in the workflow itself. The signal macro
+    // will wrap the user's function with code that performs deserialization.
+    handler: Box<dyn FnMut(&mut WF, Payload) -> Result<(), anyhow::Error>>,
+}
+pub struct QueryDefinition<WF: Workflow> {
+    // TODO: Could be a matching predicate, to allow for dynamic registration
+    name: String,
+    // The query macro will wrap the user's function with code that performs deserialization of
+    // input and serialization of output, as well as error boxing.
+    handler: Box<dyn FnMut(&WF, Payload) -> Result<Payload, anyhow::Error>>,
+}
+
+/// TODO: Placeholders, likely belong inside protos crate. These will be auto-implemented for
+///   anything using serde already (which I expect is how virtually everyone will do this).
+pub trait TemporalSerializable {}
+impl<T> TemporalSerializable for T {}
+pub trait TemporalDeserializable {}
+impl<T> TemporalDeserializable for T {}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use futures::FutureExt;
+    use std::{collections::HashMap, marker::PhantomData};
+
+    // Workflow implementation example
+    struct MyWorkflow {
+        foo: u64,
+        bar: HashMap<String, u64>,
+    }
+
+    impl Workflow for MyWorkflow {
+        type Input = String;
+        type Output = u64;
+        const NAME: &'static str = "MyWorkflowType";
+
+        fn new(input: Self::Input, _ctx: SafeWfContext) -> Self {
+            let mut bar = HashMap::new();
+            bar.insert(input, 10);
+            Self { foo: 0, bar }
+        }
+
+        fn run(
+            &mut self,
+            ctx: WfContext,
```

Comment:
When you do get around to proposal time, I think it'd be worth a discussion of the tradeoff of making this statically accessible via async-local or similar. What I've found in the last two SDKs is that I can just grab the current coroutine scheduler.

Calls like https://docs.rs/tokio/latest/tokio/task/fn.spawn.html#panics are ok assuming the runtime, I wonder if we can be too. But also I could understand from a function-grouping perspective that having it on a context may be clearer.

One benefit of not having a manually propagated context is that users aren't tempted to store it to call from a query handler or something.

Hrmm, now I wonder if we can have the trait have `fn context(&mut self) -> WorkflowContext` and `fn safe_context(&self) -> SafeWorkflowContext` on it. But that doesn't help the `new` call though.

Reply:
We could. I would be OK with it I think. This feels somewhat more obvious to me, and I like that you don't have these error paths for "called in some context where it doesn't make sense".